primitives/src/pcs/multilinear_kzg/srs.rs

```diff
@@ -123,13 +117,148 @@ impl<E: Pairing> StructuredReferenceString for MultilinearUniversalParams<E> {
+    #[cfg(any(test, feature = "test-srs"))]
+    fn gen_srs_for_testing<R>(rng: &mut R, num_vars: usize) -> Result<Self, PCSError>
+    where
+        R: ark_std::rand::RngCore + ark_std::rand::CryptoRng,
+    {
+        tests::gen_srs_for_testing(rng, num_vars)
+    }
+}
+
+// Implement `trait StructuredReferenceString` for (ML_pp, Uni_pp) to be used in
+// MLE PCS.
+impl<E: Pairing> StructuredReferenceString
+    for (MultilinearUniversalParams<E>, UnivariateUniversalParams<E>)
+{
+    type ProverParam = (MultilinearProverParam<E>, UnivariateProverParam<E>);
+    type VerifierParam = (MultilinearVerifierParam<E>, UnivariateVerifierParam<E>);
+
+    fn trim(
+        &self,
+        supported_size: usize,
+    ) -> Result<(Self::ProverParam, Self::VerifierParam), PCSError> {
+        let ml_pp = <MultilinearUniversalParams<E> as StructuredReferenceString>::trim(
+            &self.0,
+            supported_size,
+        )?;
+        let uni_pp = <UnivariateUniversalParams<E> as StructuredReferenceString>::trim(
+            &self.1,
+            supported_size,
+        )?;
+
+        Ok(((ml_pp.0, uni_pp.0), (ml_pp.1, uni_pp.1)))
+    }
+
+    fn extract_prover_param(&self, supported_size: usize) -> Self::ProverParam {
+        let ml_prover_param =
+            <MultilinearUniversalParams<E> as StructuredReferenceString>::extract_prover_param(
+                &self.0,
+                supported_size,
+            );
+        let uni_prover_param =
+            <UnivariateUniversalParams<E> as StructuredReferenceString>::extract_prover_param(
+                &self.1,
+                supported_size,
+            );
+
+        (ml_prover_param, uni_prover_param)
+    }
+
+    fn extract_verifier_param(&self, supported_size: usize) -> Self::VerifierParam {
+        let ml_verifier_param =
+            <MultilinearUniversalParams<E> as StructuredReferenceString>::extract_verifier_param(
+                &self.0,
+                supported_size,
+            );
+        let uni_verifier_param =
+            <UnivariateUniversalParams<E> as StructuredReferenceString>::extract_verifier_param(
+                &self.1,
+                supported_size,
+            );
+
+        (ml_verifier_param, uni_verifier_param)
+    }
+
+    #[cfg(any(test, feature = "test-srs"))]
+    fn gen_srs_for_testing<R>(rng: &mut R, supported_size: usize) -> Result<Self, PCSError>
+    where
+        R: ark_std::rand::RngCore + ark_std::rand::CryptoRng,
+    {
+        let ml_pp =
+            <MultilinearUniversalParams<E> as StructuredReferenceString>::gen_srs_for_testing(
+                rng,
+                supported_size,
+            )?;
+        let uni_pp =
+            <UnivariateUniversalParams<E> as StructuredReferenceString>::gen_srs_for_testing(
+                rng,
+                supported_size,
+            )?;
+        Ok((ml_pp, uni_pp))
+    }
+}
+
+#[cfg(any(test, feature = "test-srs"))]
+mod tests {
```

Comment:
Is this some new test code? Do I need to review this part?

Reply:
nope, I just moved this code over, no code changes, just reorg.